### PR TITLE
feat(Leo): Support editing AI replies

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/35343

- Allow modifying AI replies when it's not generating.
- Add a new edited label based on figma update.

Screenshot 2024-07-31 at 2 00 27 PM
Screenshot 2024-07-31 at 2 06 56 PM

https://www.figma.com/design/m0Gdbf0wtqyfEFGm32VLLc/%F0%9F%94%84-Leo-%5BIN-PROGRESS%5D?node-id=4604-59575&m=dev

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Edit the AI replies in Leo panel and continue the conversation, the edited reply will be used in further conversation.
Also test edit the user prompt to check there's no regression.

### CODE REVIEW

File: components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc [hunk 1]
```diff
@@ -124,7 +124,12 @@ void EngineConsumerConversationAPI::GenerateAssistantResponse(
     }
     ConversationEvent event;
     event.role = message->character_type;
-    event.content = (message == last_turn) ? human_input : message->text;
+
+    const std::string& text = (message->edits && !message->edits->empty())
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/engine/engine_consumer_conversation_api.cc:128]: Optional:

I wonder if instead of having `edits` and `text` `text` should just return whatever is on the top of the edit stack?

Then we wouldn't need to do `edits.empty() ? text : edits.back()` every time

Optional 'cause it'd be quite a change in this PR and is a bit unrelated :smile:

File: components/ai_chat/core/browser/engine/engine_consumer_oai.cc [hunk 2]
```diff
@@ -88,6 +88,9 @@ base::Value::List BuildMessages(
     message.Set("role", turn->character_type == CharacterType::HUMAN
                             ? "user"
                             : "assistant");
+    const std::string& text = (turn->edits && !turn->edits->empty())
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/engine/engine_consumer_oai.cc:91]: In other places we don't test `turn->edits`, we just check whether its empty? Do we need to update the other places to do this check too?

In the mojom, it seems to be nullable, so maybe we should? Alternatively, we could just make it a required field (and set it to the empty array)

File: components/ai_chat/core/browser/conversation_driver.cc [hunk 3]
```diff
@@ -1350,10 +1350,60 @@ void ConversationDriver::ModifyConversation(uint32_t turn_index,
...
+    auto edited_turn = mojom::ConversationTurn::New(
+        turn->character_type, turn->action_type, turn->visibility,
+        trimmed_input, std::nullopt /* selected_text */, std::move(events),
+        base::Time::Now(), std::nullopt /* edits */);
+    edited_turn->events->at(*completion_event_index)
+        ->get_completion_event()
+        ->completion = trimmed_input;
+
+    if (!turn->edits) {
+      turn->edits.emplace();
```
Comment (fallaciousreasoning) [inline components/ai_chat/core/browser/conversation_driver.cc:1394]: Wow, didn't know you could construct mojom structs like that! Cool!

File: components/ai_chat/resources/page/components/conversation_list/index.tsx [hunk 4]
```diff
@@ -114,8 +119,11 @@ function ConversationList(props: ConversationListProps) {
             id === 1 && isAIAssistant && context.shouldShowLongPageWarning
           const showEditInput = editInputId === id
           const showEditIndicator = !showEditInput && !!turn.edits?.length
-          const latestEdit = turn.edits?.at(-1);
-          const latestTurnText = latestEdit?.text ?? turn.text
+          const latestEdit =
+            (turn.edits && turn.edits.length > 0) ? turn.edits.at(-1) : null
```
Comment (fallaciousreasoning) [inline components/ai_chat/resources/page/components/conversation_list/index.tsx:123]:
```suggestion
          const latestEdit =
            turn.edits?.at(-1)
```

**Review by fallaciousreasoning — APPROVED** [review]

**Review by fallaciousreasoning — APPROVED** [review]
LGTM % a few (mostly optional) nits!

File: components/ai_chat/core/browser/engine/engine_consumer_oai.cc (hunk 2 above)
Comment (yrliou) [reply components/ai_chat/core/browser/engine/engine_consumer_oai.cc:91]: Sorry would you mind point me where we only check it's empty in c++?

File: components/ai_chat/resources/page/components/conversation_list/index.tsx (hunk 4 above)
Comment (yrliou) [reply components/ai_chat/resources/page/components/conversation_list/index.tsx:123]: I thought it can't handle empty array, but I think I was wrong, updating! :)

Comment (yrliou) [reply components/ai_chat/resources/page/components/conversation_list/index.tsx]: updated

File: components/ai_chat/core/browser/engine/engine_consumer_oai.cc (hunk 2 above)
Comment (fallaciousreasoning) [reply components/ai_chat/core/browser/engine/engine_consumer_oai.cc:91]: Sorry I think I must've hallucinated this - I had a look but can't seem to find it! Sorry!
